### PR TITLE
Fix reward_processing_classes validation in GRPOTrainer (#2839)

### LINKED ISSUE #2839: A bug in grpo_trainer.py
### Reproduction

**Issue Description:**

I’m reading the original code in the file grpo_trainer.py and found an issue related to the matching of reward_processing_classes and reward_funcs. The problem occurs when reward_processing_classes is a single object, which gets transformed into a list, but no length check is performed to ensure it matches the number of reward_funcs. This leads to unexpected behavior when there are multiple reward_funcs.

**Code in question:**
The issue appears in lines 225 to 234 in grpo_trainer.py:
```python

        # Reward processing class
        if reward_processing_classes is None:
            reward_processing_classes = [None] * len(reward_funcs)
        elif not isinstance(reward_processing_classes, list):
            reward_processing_classes = [reward_processing_classes]
        else:
            if len(reward_processing_classes) != len(reward_funcs):
                raise ValueError("The number of reward processing classes must match the number of reward functions.")

        for i, (reward_processing_class, reward_func) in enumerate(zip(reward_processing_classes, reward_funcs)):
              ...

```
**Bug Explanation:**

If reward_processing_classes is a pretrained object (not a list), it gets converted into a list with one element. However, no check is made to ensure that the number of elements in reward_processing_classes matches the number of reward_funcs.
The following code(line 234):
```
        for i, (reward_processing_class, reward_func) in enumerate(zip(reward_processing_classes, reward_funcs)):
```
will zip reward_processing_classes (which now has only one element) with reward_funcs. As a result, if len(reward_funcs) > 1, only the first reward_func will be processed, and the rest will be discarded.

**Expected Behavior:**
If reward_processing_classes is provided as a single object (not a list), it should automatically be expanded to match the number of reward_funcs.
If the length of reward_processing_classes is 1, the code should duplicate the object to match the length of reward_funcs.

I hope this issue can be addressed soon.

### System Info

trl = 0.14.0

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [ ] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

### PR DESCRIPTION
**Pull request by chi2liu**
## What does this PR do?

Fixes #2839 - When a single `reward_processing_class` is provided for multiple reward functions in GRPOTrainer, it now correctly replicates the processing class for each reward function instead of silently ignoring all but the first reward function.

## Bug description

Previously, when users provided:
- Multiple reward functions (e.g., `[func1, func2, func3]`)
- A single reward processing class (not a list)

The code would convert the single processing class to a list with one element `[processing_class]`, then zip it with the reward functions. This resulted in only the first reward function being properly processed, while the rest were silently ignored.

## Solution

The fix now replicates the single processing class to match the number of reward functions:
```python
reward_processing_classes = [reward_processing_classes] * len(reward_funcs)
```

This ensures all reward functions get the same processing class, which is the expected behavior when providing a single processing class.

## Testing

Added comprehensive unit test `test_reward_processing_classes_replication` that verifies:
1. Single processing class is correctly replicated for multiple reward functions
2. List validation still works (mismatched lengths raise ValueError)
3. Correct list length continues to work as expected

All existing tests pass without regression.

## Before submitting
- [x] This PR fixes an issue
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request)
- [x] Did you make sure to update the documentation with your changes? (No doc changes needed)
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed.

### CODE REVIEW

**Comment by qgallouedec** [comment]
We don't expect a different number of elements in reward_funcs and reward_processing_class. I think we should raise and exception if this case occurs

**Comment by chi2liu** [comment]
> We don't expect a different number of elements in reward_funcs and reward_processing_class. I think we should raise and exception if this case occurs

sure, updated it

File: tests/test_grpo_trainer.py [hunk 1]
```diff
@@ -1907,6 +1907,84 @@ def test_training_sequence_importance_sampling(self):
...
+        reward_funcs = [reward_func1, reward_func2]
+
+        # Create a single processing class (tokenizer)
+        single_processing_class = AutoTokenizer.from_pretrained("gpt2")
+        single_processing_class.pad_token = single_processing_class.eos_token
+
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            training_args = GRPOConfig(
+                output_dir=tmp_dir,
+                max_steps=1,
```
Comment (qgallouedec) [inline tests/test_grpo_trainer.py]:
```suggestion
```

File: tests/test_grpo_trainer.py [hunk 2]
```diff
@@ -1907,6 +1907,84 @@ def test_training_sequence_importance_sampling(self):
                 new_param = trainer.model.get_parameter(n)
                 self.assertFalse(torch.equal(param, new_param), f"Parameter {n} has not changed.")
 
+    def test_reward_processing_classes_validation(self):
```
Comment (qgallouedec) [inline tests/test_grpo_trainer.py]: Il not quite sure about this test. I think it would make more sense to pass two reward **models** (not function) and one reward processing class

File: tests/test_grpo_trainer.py [hunk 3]
```diff
@@ -1907,6 +1907,84 @@ def test_training_sequence_importance_sampling(self):
...
+                    args=training_args,
+                    train_dataset=dataset,
+                )
+
+            self.assertIn("must match", str(context.exception))
+
+            # Test 3: Correct list length should work
+            correct_processing_classes = [single_processing_class, single_processing_class]
+
+            trainer2 = GRPOTrainer(
```
Comment (qgallouedec) [inline tests/test_grpo_trainer.py]: Let's try to avoid two instance of trainer in one test case. If you want to test two cases, best is to have two tests

File: tests/test_grpo_trainer.py (hunk 1 above)
Comment (chi2liu) [reply tests/test_grpo_trainer.py]: updated

File: tests/test_grpo_trainer.py (hunk 3 above)
Comment (chi2liu) [reply tests/test_grpo_trainer.py]: updated

File: tests/test_grpo_trainer.py (hunk 2 above)
Comment (chi2liu) [reply tests/test_grpo_trainer.py]: updated

File: trl/trainer/grpo_trainer.py [hunk 4]
```diff
@@ -609,10 +609,20 @@ def __init__(
...
+                    f"Received a single reward processing class instead."
+                )
             reward_processing_classes = [reward_processing_classes]
         else:
             if len(reward_processing_classes) != len(reward_funcs):
-                raise ValueError("The number of reward processing classes must match the number of reward functions.")
+                raise ValueError(
+                    f"The number of reward processing classes ({len(reward_processing_classes)}) must match "
+                    f"the number of reward functions ({len(reward_funcs)})."
+                )
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py]:
```suggestion
        elif not isinstance(reward_processing_classes, list):
            reward_processing_classes = [reward_processing_classes]
        if len(reward_processing_classes) != len(reward_funcs):
            raise ValueError(
                f"The number of reward processing classes ({len(reward_processing_classes)}) must match the number of "
                f"reward functions ({len(reward_funcs)})."
            )
```

maybe this is simpler?

File: tests/test_grpo_trainer.py [hunk 5]
```diff
@@ -1907,6 +1907,136 @@ def test_training_sequence_importance_sampling(self):
...
+
+        # Use two reward models
+        reward_models = [
+            "trl-internal-testing/tiny-Qwen2ForSequenceClassification-2.5",
+            "trl-internal-testing/tiny-Qwen2ForSequenceClassification-2.5",
+        ]
+
+        # Create a single processing class (tokenizer)
+        single_processing_class = AutoTokenizer.from_pretrained("gpt2")
+        single_processing_class.pad_token = single_processing_class.eos_token
```
Comment (qgallouedec) [inline tests/test_grpo_trainer.py]: why using gpt2 here? just use the same model (`"trl-internal-testing/tiny-Qwen2ForSequenceClassification-2.5"`)

File: trl/trainer/grpo_trainer.py (hunk 4 above)
Comment (chi2liu) [reply trl/trainer/grpo_trainer.py]: updated

File: tests/test_grpo_trainer.py (hunk 5 above)
Comment (chi2liu) [reply tests/test_grpo_trainer.py]: updated

**Review by qgallouedec — APPROVED** [review]
thanks!
